File: app/src/main/java/com/example/coloreffect/NoteDataSource.java
package com.example.coloreffect;

import android.content.ContentValues;
import android.content.Context;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;

import java.io.Closeable;
import java.io.IOException;

//  Источник данных, позволяет изменять данные в таблице
// Создает и держит в себе читатель данных
public class NoteDataSource implements Closeable {
    private DatabaseHelper dbHelper;
    private SQLiteDatabase database;
    private NoteDataReader noteDataReader;

    public NoteDataSource(Context context) {
        dbHelper = new DatabaseHelper(context);
    }

    // Открывает базу данных
    public void open() throws SQLException {
        database = dbHelper.getWritableDatabase();
        // Создать читателя и открыть его
        noteDataReader = new NoteDataReader(database);
        noteDataReader.open();
    }

    @Override
    public void close() throws IOException {
        noteDataReader.close();
        dbHelper.close();
    }

    // Добавить новую запись
    public CityNote addNote(String title, String description) {
        ContentValues values = new ContentValues();
        values.put(DatabaseHelper.COLUMN_NOTE,description);
        values.put(DatabaseHelper.COLUMN_NOTE_TITLE,title);
        // Добавление записи
        long insertId = database.insert(DatabaseHelper.TABLE_NOTES,null,values);
        CityNote note = new CityNote();
        note.setId(insertId);
        note.setTitle(title);
        note.setDescription(description);
        return note;
    }

    // Изменить запись
    public void editNote(CityNote note, String title, String description) {
        ContentValues editedNote = new ContentValues();
        editedNote.put(DatabaseHelper.COLUMN_ID, note.getId());
        editedNote.put(DatabaseHelper.COLUMN_NOTE, description);
        editedNote.put(DatabaseHelper.COLUMN_NOTE_TITLE,title);
        // Изменение записи
        database.update(DatabaseHelper.TABLE_NOTES,
                editedNote,
                DatabaseHelper.COLUMN_ID + "=" + note.getId(),
                null);
    }

    // Удалить запись
    public void deleteNote(CityNote note) {
        long id = note.getId();
        database.delete(DatabaseHelper.TABLE_NOTES,
                DatabaseHelper.COLUMN_ID + "=" + id,
                null);
    }

    // Очистить таблицу
    public void deleteAll() {
        database.delete(DatabaseHelper.TABLE_NOTES, null, null);
    }

    // Вернуть читателя (он потребуется в других местах)
    public NoteDataReader getNoteDataReader(){
        return noteDataReader;
    }

}

File: app/src/main/java/com/example/coloreffect/NoteDataSource.kt
package com.example.coloreffect

//import com.example.coloreffect.Controller.start
import android.content.*
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import java.io.Closeable
import java.io.IOException

//  Источник данных, позволяет изменять данные в таблице
// Создает и держит в себе читатель данных
class NoteDataSource(context: Context?) : Closeable {
    private val dbHelper: DatabaseHelper
    private var database: SQLiteDatabase? = null

    // Вернуть читателя (он потребуется в других местах)
    var noteDataReader: NoteDataReader? = null
        private set

    // Открывает базу данных
    @Throws(SQLException::class)
    fun open() {
        database = dbHelper.writableDatabase
        // Создать читателя и открыть его
        noteDataReader = NoteDataReader(database)
        noteDataReader!!.open()
    }

    @Throws(IOException::class)
    override fun close() {
        noteDataReader!!.close()
        dbHelper.close()
    }

    // Добавить новую запись
    fun addNote(title: String?, description: String?): CityNote {
        val values = ContentValues()
        values.put(DatabaseHelper.Companion.COLUMN_NOTE, description)
        values.put(DatabaseHelper.Companion.COLUMN_NOTE_TITLE, title)
        // Добавление записи
        val insertId = database!!.insert(DatabaseHelper.Companion.TABLE_NOTES, null, values)
        val note = CityNote()
        note.id = insertId
        note.title = title
        note.description = description
        return note
    }

    // Изменить запись
    fun editNote(note: CityNote?, title: String?, description: String?) {
        val editedNote = ContentValues()
        editedNote.put(DatabaseHelper.Companion.COLUMN_ID, note!!.id)
        editedNote.put(DatabaseHelper.Companion.COLUMN_NOTE, description)
        editedNote.put(DatabaseHelper.Companion.COLUMN_NOTE_TITLE, title)
        // Изменение записи
        database!!.update(
            DatabaseHelper.Companion.TABLE_NOTES,
            editedNote,
            DatabaseHelper.Companion.COLUMN_ID + "=" + note.id,
            null
        )
    }

    // Удалить запись
    fun deleteNote(note: CityNote?) {
        val id = note!!.id
        database!!.delete(
            DatabaseHelper.Companion.TABLE_NOTES,
            DatabaseHelper.Companion.COLUMN_ID + "=" + id,
            null
        )
    }

    // Очистить таблицу
    fun deleteAll() {
        database!!.delete(DatabaseHelper.Companion.TABLE_NOTES, null, null)
    }

    init {
        dbHelper = DatabaseHelper(context)
    }
}